onnxruntime/core/platform/posix/env.cc

```diff
@@ -262,16 +271,51 @@ class PosixEnv : public Env {
     return new PosixThread(name_prefix, index, start_address, param, thread_options);
   }
 
+  // Return the number of physical cores
   int GetNumCpuCores() const override {
-    // TODO if you need the number of physical cores you'll need to parse
-    // /proc/cpuinfo and grep for "cpu cores".
-    // However, that information is not always available(output of 'grep -i core /proc/cpuinfo' is empty)
-    return std::thread::hardware_concurrency();
+#ifdef CPUINFO_SUPPORTED
+    if(cpuinfo_available_) {
+      return gsl::narrow<int>(cpuinfo_get_cores_count());
+    }
+#endif
+    // We guess the number of cores
+    return std::max(1, static_cast<int>(std::thread::hardware_concurrency() / 2));
   }
 
-  std::vector<size_t> GetThreadAffinityMasks() const override {
-    std::vector<size_t> ret(std::thread::hardware_concurrency() / 2);
-    std::iota(ret.begin(), ret.end(), 0);
+  std::vector<LogicalProcessors> GetThreadAffinityMasks() const override {
+
+    std::vector<LogicalProcessors> ret;
+#ifdef CPUINFO_SUPPORTED
+    if (cpuinfo_available_) {
+#if !defined(__APPLE__) && !defined(__ANDROID__) && !defined(__wasm__) && !defined(_AIX)
+      // We currently do not implement affinity on more than 64 cores.
+      auto num_phys_cores = cpuinfo_get_cores_count();
+      ret.reserve(num_phys_cores);
+      for (uint32_t i = 0; i < num_phys_cores; ++i) {
+        const auto* core = cpuinfo_get_core(i);
+        LogicalProcessors th_aff;
+        // Processor count will never exceed 64 in a given group.
+        // TBD: Processor groups are currently not taken into account.
+        th_aff.reserve(core->processor_count);
+        auto log_proc_idx = core->processor_start;
+        for (uint32_t count = 0; count < core->processor_count; count++, ++log_proc_idx) {
+          const auto* log_proc = cpuinfo_get_processor(log_proc_idx);
+          th_aff.push_back(log_proc->linux_id);
+        }
+        ret.push_back(std::move(th_aff));
+       }
+#else
+      // We do not set affinity on these platforms, so we just generate an vector sized for cores
+      ret.resize(cpuinfo_get_cores_count());
+#endif
+    }
+#endif // CPUINFO_SUPPORTED
+    // We do not know what the affinities should be, so we just guess that HT is 2 logical
+    // processors per physical cores, but do not guess the affinities as we do not know the
+    // actual processor IDs.
+    if(ret.empty()) {
+      ret.resize(std::thread::hardware_concurrency()/2);
```

Comment:
>ret.resize(std::thread::hardware_concurrency()/2); [](http://example.com/codeflow?start=6&length=50)

the way we handle this empty() case varies among windows and posix - windows will throw excepiton while posix will set a default by a std call.
we should align the two by letting windows do the same instead of throwing.

Reply:
The consensus was to handle it with hardware_concurrency/2.